devtools/x/src/cargo.rs

```diff
@@ -332,14 +332,39 @@ impl<'a> CargoCommand<'a> {
             return Ok(());
         }
 
+        let mut cargo = self.prepare_cargo(packages);
+        cargo.run()
+    }
+
+    /// Runs this command on the selected packages, returning the standard output as a bytestring.
+    pub fn run_capture_messages(
+        &self,
+        packages: &SelectedPackages<'_>,
+    ) -> Result<impl Iterator<Item = Result<Message>>> {
+        // Early return if we have no packages to run.
+        let output = if !packages.should_invoke() {
+            info!("no packages to {}: exiting early", self.as_str());
+            vec![]
+        } else {
+            let mut cargo = self.prepare_cargo(packages);
+            cargo.args(&["--message-format", "json"]);
+            cargo.run_with_output()?
+        };
+
+        Ok(Message::parse_stream(Cursor::new(output))
+            .map(|message| message.context("error while parsing message from Cargo")))
+    }
+
+    fn prepare_cargo(&self, packages: &SelectedPackages<'_>) -> Cargo {
         let mut cargo = Cargo::new(self.cargo_config(), self.as_str(), true);
         cargo
             .current_dir(project_root())
             .args(self.direct_args())
             .packages(packages)
             .pass_through(self.pass_through_args())
-            .envs(self.get_extra_env().to_owned())
-            .run()
```

Comment:
I guess waiting for run to be invoked via asking for output is better given this functions name.    Just dislike that we implicitly call run() as it's a rather important call to invoke implicitly.

Reply:
Hmm, I think the refactoring I did here might be slightly confusing: we do call `run_with_output()` which has `run` in the name, so hopefully that's explicit enough...